### PR TITLE
perf(huff0): #167 cheap entropy proxy for table_log selection — no FSE-encode per candidate

### LINKED ISSUE #167: perf(huff0): donor HUF_optimalTableLog fast-path single-shot tableLog with cheap description-size proxy
## Context

Surfaced during review of #166 (`FSE_buildCTable_wksp` parity / drop per-symbol `Vec<State>`). Part of #111 Phase 7 Lane A — block compressor entropy coding speed.

Profile of `compress/level_2_dfast/small-4k-log-lines/matrix/pure_rust` (post-#166):

| % inclusive | Function |
|---|---|
| ~32 % | `huff0_encoder::HuffmanTable::try_table_description_size` |
| ~23 % | `huff0_encoder::HuffmanTable::build_from_counts` |
| ~18 % | `huff0_encoder::HuffmanEncoder::encode_weight_description` |

## Problem

`HuffmanTable::build_from_counts` runs a speculative search over `min_table_log..=11`, calling `try_table_description_size` on each candidate. `try_table_description_size` actually **FSE-encodes the weights** to a fresh buffer — this is the expensive operation. The loop pays 7× FSE-encode-of-weights to pick the optimal `(desc_size + payload_size)` `tableLog`.

Donor's `HUF_optimalTableLog` (`lib/compress/huf_compress.c:1272`) has two modes:

1. **Default (fast)** — pure arithmetic via `FSE_optimalTableLog_internal(maxLog=11, srcSize, maxSV, minus=1)`. Single `tableLog`, no per-candidate rebuild.
2. **`HUF_flags_optimalDepth`** — full search loop. Gated by `strategy >= ZSTD_btultra` (level 18+).

For all strategies below btultra (every dfast/lazy level in this codebase's hot path) donor uses the fast path. Rust currently does the slow search unconditionally.

## What was tried and why it was reverted

Implemented the donor fast path inside #166 work (Stage 1). Result on `compress/level_2_dfast/small-4k-log-lines/matrix/pure_rust`:

- Speed: −22 % on top of the #166 −43 %.
- Ratio: **regressed vs donor on one cell** — `small-4k-log-lines/L1_fast` `rust_bytes=159` vs `ffi_bytes=157` (+2 B). Pre-#166 search loop produced `R=154` (better than donor); single-shot loses that 5 B margin and falls 2 B below the donor.

Per project rule \"Ratio first — if `rust_bytes > ffi_bytes` we lose vs donor → real bug\", the single-shot change cannot land while it produces a `rust_bytes > ffi_bytes` cell. Stage 1 reverted; only Stage 2 (PR #166) landed.

## Suggested approach

The expensive component of the current loop is `try_table_description_size`, which FSE-encodes weights to count bytes. Replace that with a **cheap description-size proxy** that approximates the FSE-encoded weight description size from a histogram of the per-symbol weights (no actual FSE encode). The proxy should:

1. Compute the histogram of weight values (13 bins, donor max weight = 12).
2. Approximate the FSE-compressed weight description size analytically — e.g. via `H(weights) * total_weights / 8` plus a small constant for the FSE header. Possibly compose with a min-of-raw-encoding (`weights.len().div_ceil(2) + 1`) since the donor picks the cheaper of FSE / raw.
3. Validate that the proxy's relative ranking across `tableLog` candidates matches `try_table_description_size` ranking on the bench corpus (small-4k-log-lines, decodecorpus-z000033, large-log-stream, high-entropy-1m).
4. Verify ratio is **at least as good as the current search loop** vs the donor on every level × scenario cell (compare_ffi REPORT sweep).

If the proxy preserves the optimum within tolerance, the loop becomes:

```rust
for table_log in min_table_log..=11 {
    let weights = build_donor_limited_weights(counts, table_log);
    if !huffman_weight_sum_is_power_of_two(&weights) { continue; }
    let table = Self::build_from_weights(&weights);
    let max_bits = table.codes.iter().map(|&(_, b)| b).max().unwrap_or(0) as usize;
    if max_bits < table_log && table_log > min_table_log { break; }
    let desc_size_est = huff_desc_size_proxy(&weights); // <- cheap
    let new_size = table.estimate_compressed_size_from_counts(counts) + desc_size_est;
    if new_size < best_size { best_size = new_size; best_table = Some(table); }
    if new_size > best_size + 1 { break; }
}
```

— same selection shape, no FSE-encode per candidate. Expected speedup matches the Stage 1 measurement (~−22 % on small-4k-log L2_dfast on top of #166), with no ratio regression vs donor since the proxy still drives the same selection criterion.

## Files involved

- `zstd/src/huff0/huff0_encoder.rs` — `HuffmanTable::build_from_counts` (~line 220), `HuffmanEncoder::encode_weight_description` (~line 125), new proxy fn.
- `zstd/benches/compare_ffi.rs` — `REPORT` ratio sweep is the validation gate.

## Acceptance criteria

- [ ] Cheap description-size proxy lands in `huff0_encoder.rs` with `debug_assert` cross-checks against `try_table_description_size` on a representative sample (e.g. property test).
- [ ] `compare_ffi --features dict_builder` REPORT sweep: **zero new cells** where `rust_bytes > ffi_bytes` compared to current main.
- [ ] `compress/level_2_dfast/small-4k-log-lines/matrix/pure_rust` improves at least −10 % vs current main (post-#166 47.4 µs baseline).
- [ ] 501/501 lib tests pass, clippy clean.

## Related

- Part of #111 Phase 7 Lane A (block compressor entropy coding speed).
- Builds on #166 (FSE_buildCTable_wksp parity, drop per-symbol Vec<State>).
- Builds on #164 / PR #165 (FSE encode flat nextStateTable + symbolTT).
- Surfaced 2026-05-18 during /review-threads on #166.

### PR DESCRIPTION
**Pull request by polaz**
## Summary

Closes #167. `HuffmanTable::build_from_counts` previously ran a `min_table_log..=11` speculative search, calling `try_table_description_size` per candidate — which actually **FSE-encodes the weight stream against a freshly-built FSE table** just to count bytes. That was ~31% inclusive on the post-#166 4 KiB compress profile.

Replace with `cheap_desc_size_proxy(weights)`: an integer entropy estimate that reproduces the donor `HUF_writeCTable_wksp` decision (FSE vs raw nibble) **without touching the FSE encoder**.

## Algorithm

For each `table_log` candidate:
- **Raw nibble** — exact size = `weights.len().div_ceil(2) + 1`. Representable when `weights.len() <= 128`.
- **FSE** — `sum(c_i * ceil_log2(total / c_i))` bits over the 13-bin weight histogram + `8 B` header constant. Empirically-derived upper bound for the `acc_log = 6` weight FSE table. Representable when result < 128 B.
- Return `min(fse_size, raw_size)` when both representable.

Donor `HUF_writeCTable_wksp` picks the smaller of the two; the proxy mirrors that decision analytically.

## Validation

**Ratio sweep** (`STRUCTURED_ZSTD_EMIT_REPORT=1 cargo bench --bench compare_ffi --features dict_builder`) — every scenario × every supported level:

- **Zero new `rust_bytes > ffi_bytes` cells** introduced.
- `small-4k-log-lines/L1_fast` preserved at `R=154` (≤ `C=157`) — the cell that broke when the donor `HUF_optimalTableLog` fast-path single-shot was tried in #166 review and reverted.
- `decodecorpus-z000033/level_18_btultra` + `/level_19_btultra` **improved** by 442 B (R=443 434 → 442 992) — the proxy steers selection to a marginally tighter `table_log` on those cells.
- Every other R/C pair: unchanged or improved.

**Speed** — `compress/level_2_dfast/small-4k-log-lines/matrix/pure_rust`:

| Baseline | Time | vs pre-#165 |
|----------|------|-------------|
| pre-#165 | 83.7 µs | — |
| #165 (FSE encode_sequences O(1)) | 59.7 µs | −29 % |
| #166 (FSE build donor parity) | 47.4 µs | −43 % |
| **this PR (#167 cheap proxy)** | **34.2 µs** | **−59 %** (+145 % thrpt) |

Gap vs C donor (7.1 µs): 11.8× → 4.8×.

No regression on larger corpora (entropy table-log search wasn't the bottleneck on z000033 — 13.1 ms unchanged).

## Test plan
- 502/502 lib tests pass via `cargo nextest`, including new `cheap_desc_size_proxy_is_conservative_vs_exact` sanity test.
- `cargo clippy --all-targets -- -D warnings` clean.
- `cargo fmt --check` clean.
- `compare_ffi` REPORT sweep documented above.

Closes #167.

## Summary by CodeRabbit

* **Refactor**
  * Improved Huffman table selection to speed candidate evaluation, reduce transient memory use, and produce more consistent/compressed table descriptions for better overall compression efficiency.

* **Tests**
  * Added unit tests covering varied weight distributions and edge cases to validate description-size estimation, representability decisions, and selection behavior.

[Review Change Stack](https://app.coderabbit.ai/change-stack/structured-world/structured-zstd/pull/168?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: zstd/src/huff0/huff0_encoder.rs [hunk 1]
```diff
@@ -908,6 +1035,119 @@ fn from_data() {
...
+            (Some(p), Some(e)) => {
+                // Proxy is allowed to overestimate; flag a hard
+                // underestimate vs raw (the natural floor — if the
+                // exact path picks raw nibble, the proxy must not be
+                // below `n.div_ceil(2) + 1 - 1`).
+                let raw_floor = weights.len().div_ceil(2);
+                assert!(
+                    p + 2 >= e || p >= raw_floor,
+                    "[{label}] proxy {p} under-shot exact {e} (raw_floor {raw_floor})"
+                );
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/huff0/huff0_encoder.rs]: @coderabbitai Fixed — rebuilt fixtures from `HuffmanTable::build_from_counts(counts)` so every weight vector is Kraft-valid by construction (the encoder's own output passes its own gate). Added `exercised > 0` assertion to catch a future regression silently making the test vacuous again. `raw_floor` now computed from `trimmed.len().div_ceil(2) + 1` per the suggestion.
